<a href="https://colab.research.google.com/github/shanmughamkarthik-eng/python_practice/blob/main/loops_and_control_structures_assignments_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem 6.8: Simple Calculator (Medium)

Create a simple calculator program that:

1. Shows a menu of operations: Add, Subtract, Multiply, Divide, Exit
2. Takes two numbers as input
3. Performs the selected operation
4. Shows the result
5. Repeats until the user chooses to exit
6. Handles division by zero error

Hint: Use while loop. Display menu, get choice. Use if-elif to perform operation. Check if divisor is 0 before division.

In [ ]:
class calculator:
  def __init__(self, number_a = 0, number_b = 0):
    self.number_a = number_a
    self.number_b = number_b

  def addition(self):
    return self.number_a + self.number_b

  def subtraction(self):
    return self.number_a - self.number_b

  def multiplication(self):
      return self.number_a * self.number_b

  def division(self):
      return self.number_a / self.number_b if self.number_b != 0 else "Error: Division by zero"



print("Simple Calculator")

calc = calculator()

while True:
  print("\n1. Add\n2. Subtract\n3. Multiply\n4. Divide\n5. Exit")
  choice = input("\nEnter your choice: ")

  if choice == "5":
    print("Exiting...")
    break

  elif choice in ["1", "2", "3", "4"]:
    try:
      calc.number_a = float(input("Enter first number: "))
      calc.number_b = float(input("Enter second number: "))

      if choice == "1":
        print("Result: ", calc.addition())

      elif choice == "2":
        print("Result: ", calc.subtraction())

      elif choice == "3":
        print("Result: ", calc.multiplication())

      elif choice == "4":
        print("Result: ", calc.division())

    except ValueError:
      print("Invalid input. Please enter a valid number.")

  else:
    print("Invalid choice. Please select a valid option.")


Simple Calculator

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 1
Enter first number: 20
Enter second number: 212
Result:  232.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 2
Enter first number: 23
Enter second number: 12
Result:  11.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 3
Enter first number: 12
Enter second number: 13
Result:  156.0

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 4
Enter first number: 123
Enter second number: 0
Result:  Error: Division by zero

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 123
Invalid choice. Please select a valid option.

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 4
Enter first number: 123
Enter second number: 2
Result:  61.5

1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Enter your choice: 5
Exiting...


#Section 7: Bonus Challenges
##Problem 7.1: Collatz Conjecture (Medium)
The Collatz conjecture is one of the most famous unsolved problems in mathematics. The rules are:

1. Start with any positive integer n
2. If n is even, divide it by 2
3. If n is odd, multiply it by 3 and add 1
4. Repeat until you reach 1

###Write a program that:
*  Takes a number as input
*  Shows the complete sequence until it reaches 1
*  Counts how many steps it took

Example: Starting with 6: 6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1 (8 steps)

Hint: While n != 1: if n%2==0, n=n//2, else n=3*n+1. Count steps and print each value.

In [ ]:
print("Collatz Conjecture")

number = int(input("Enter a number: "))
steps = 0
print(f"\n{number}", end="")

while number!=1:
  if number % 2 == 0:
    number = number/2

  else:
    number = (3* number)+1

  print(f" -> {int(number)}", end="")
  steps += 1

print(f" ({steps} steps)")

Collatz Conjecture
Enter a number: 11

11 -> 34 -> 17 -> 52 -> 26 -> 13 -> 40 -> 20 -> 10 -> 5 -> 16 -> 8 -> 4 -> 2 -> 1 (14 steps)


###Problem 7.2: Caesar Cipher (Hard)
Write a program that implements a Caesar cipher to encrypt and decrypt messages.

The Caesar cipher shifts each letter by a fixed number of positions in the alphabet.

Example: With shift=3
*  'A' becomes 'D'
*  'Hello' becomes 'Khoor'

Your program should:

1. Ask for a message and shift value
2. Provide options to encrypt or decrypt
3. Handle both uppercase and lowercase letters
4. Keep non-alphabetic characters unchanged

Hint: Loop through each character. Use ord() to get ASCII value, shift it, use chr() to convert back. Use modulo 26 for wrapping. Check if char.isalpha().

In [ ]:
def get_shift_value():
  while True:
    try:
      return int(input("Enter the shift value: "))
    except ValueError:
      print("Invalid value. Please enter a shift value number.")

def caesar_cipher(message, shift_value):
  result = ""
  for char in message:
    if char.isupper():
        result += chr((ord(char) + shift_value - 65) % 26 + 65)

    elif char.islower():
        result += chr((ord(char) + shift_value - 97) % 26 + 97)

    else:
        result += char
  return result

# ------------------------

print("Caesar Cipher\n")

print("1. Encrypt")
print("2. Decrypt")
print("0. Quit")

while True:
  try:
    choice = int(input("\nEnter your choice: "))
  except ValueError:
    print("Invalid input. Please enter a valid number.")
    continue

  if choice == 0:
    print("Exiting")
    break

  elif choice == 1:
    message = input("Enter the message: ")
    shift_value = get_shift_value()
    print(f"Encrypted message: {caesar_cipher(message, shift_value)}")

  elif choice == 2:
    message = input("Enter the message: ")
    shift_value = get_shift_value()
    print(f"Decrypted message: {caesar_cipher(message, -shift_value)}")

  else:
    print("Invalid choice. Please select a valid option. eg., 0, 1 or 2")

Caesar Cipher

1. Encrypt
2. Decrypt
0. Quit

Enter your choice: w
Invalid input. Please enter a valid number.

Enter your choice: 9
Invalid choice. Please select a valid option. eg., 0, 1 or 2

Enter your choice: 1
Enter the message: HelloZ@125
Enter the shift value: 8
Encrypted message: PmttwH@125

Enter your choice: 2
Enter the message: PmttwH@125
Enter the shift value: 8
Decrypted message: HelloZ@125

Enter your choice: 1
Enter the message: hello
Enter the shift value: q
Invalid value. Please enter a shift value number.
Enter the shift value: 4
Encrypted message: lipps

Enter your choice: 2
Enter the message: lipps
Enter the shift value: d
Invalid value. Please enter a shift value number.
Enter the shift value: 4
Decrypted message: hello

Enter your choice: 0
Exiting


####Problem 7.3: ATM Machine Simulator (Hard)
#####Create an ATM machine simulator with the following features:

1. Initial balance: 10000
2. PIN: 1234 (3 attempts allowed)
3. Menu options:
  *   Check Balance
  *   Withdraw Money (check for sufficient balance and valid amount)
  *   Deposit Money
  *   Change PIN
  *   Exit
4. Keep transaction history (last 5 transactions)
5. Input validation for all operations

Hint: Use variables for balance, pin. Loop for PIN attempts. Use while loop for menu. Use list to store transaction history. Validate amounts > 0 and balance >= withdrawal.

In [6]:
def withdraw_money():
  global balance
  global transaction_history

  while True:
    try:
      amount = float(input("Enter the amount: "))
    except ValueError:
      print("Please enter a numeric value.")
      continue

    if amount <= 0:
      print("Invalid amount. Please enter a positive number.")
      continue

    elif amount > balance:
      print("Insufficient balance.")
      continue

    else:
      balance -= amount
      transaction_history.append(f" - {amount}")
      print(f"Withdrawal successful. New balance: {balance}")
      return

#----------------------------------------------------------------

def deposit_money():
  global balance
  global transaction_history

  while True:
    try:
      amount = float(input("Enter the amount: "))
    except ValueError:
      print("Please enter a numeric value.")
      continue

    if amount <= 0:
      print("Invalid amount. Please enter a positive number.")
      continue

    else:
      balance += amount
      transaction_history.append(f" + {amount}")
      print(f"Deposit successful. New balance: {balance}")
      return

#-----------------------------------------------------------------

def check_balance():
  global balance
  global transaction_history

  print(f"Your balance is: {balance:,.2f}")

  if input("Do you want to see the last 5 transactions? (y/n): ").lower() == 'y':
     # This slices the list to get the last 5 items
    last_five = transaction_history[-5:]

    if not last_five:
      print("No transactions yet.")

    else:
      print("\n--- Recent Transactions ---")
      for tx in last_five:
        print(tx)
      return

  else:
    return

#-------------------------------------------------------------------

def change_pin():
  global pin
  global transaction_history
  attempts = 0

  while attempts < 3:
    try:
      current = int(input(f"Enter your current PIN (Attempt {attempts + 1}/3):"))

      if current != pin:
          print("Incorrect PIN. Authorization failed.")
          attempts += 1
          continue

      else:
        try:
          new_pin = int(input("Enter the new PIN: "))
          confirmation = int(input("Confirm the new PIN: "))

          if new_pin == confirmation:
            pin = new_pin
            print("PIN changed successfully.")
            return True

          else:
            print("Pin not matched")
            return False

        except ValueError:
          print("Please enter a numeric value.")
          continue

    except ValueError:
      print("Please enter a numeric value.")
      attempts += 1
      continue

  if attempts >= 3:
    print("Maximum number of attempts reached. Exiting.")
    return False

#-------------------------------------------------------------------------
def login():
  global pin
  attempts = 0

  while attempts < 3:
    try:
      entered_pin = int(input(f"Enter your PIN  (Attempt {attempts + 1}/3): "))
      if entered_pin == pin:
        print("Login successful.")
        return True

      else:
          print("Incorrect PIN. Authorization failed.")
          attempts += 1
          continue

    except ValueError:
      print("Please enter a numeric value.")
      attempts += 1
      continue

  return False

#-----------------------------------------------

print("ATM Machine Simulator")
balance = 10000
pin = 1234
transaction_history = []

if not login():
  print("Exiting")

else:
  while True:
    print("\n1. Check Balance")
    print("2. Withdraw Money")
    print("3. Deposit Money")
    print("4. Change PIN")
    print("5. Exit")

    try:
      choice = int(input("\nEnter your choice: "))
    except ValueError:
      print("Invalid input. Please enter a valid number.")
      continue

    if choice == 5:
      print("Exiting")
      break

    elif choice == 1:
      check_balance()

    elif choice == 2:
      withdraw_money()

    elif choice == 3:
      deposit_money()

    elif choice == 4:
      if not change_pin():
        break

    else: print("Invalid choice. Please select a valid option. eg., 1 to 5")

ATM Machine Simulator
Enter your PIN  (Attempt 1/3): 1232
Incorrect PIN. Authorization failed.
Enter your PIN  (Attempt 2/3): as
Please enter a numeric value.
Enter your PIN  (Attempt 3/3): 1234
Login successful.

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 1
Your balance is: 10,000.00
Do you want to see the last 5 transactions? (y/n): n

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 6
Invalid choice. Please select a valid option. eg., 1 to 5

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 2
Enter the amount: qw
Please enter a numeric value.
Enter the amount: 1200
Withdrawal successful. New balance: 8800.0

1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Enter your choice: 3
Enter the amount: qw
Please enter a numeric value.
Enter the amount: 3500
Deposit successful. New balance: 12300.0

1. Check Balance
2. Withdraw

    Enter your choice: 4
    Enter your current PIN (Attempt 1/3):1234
    Enter the new PIN: 1221
    Confirm the new PIN: 1222
    Pin not matched

    ATM Machine Simulator
    Enter your PIN  (Attempt 1/3): 1232
    Incorrect PIN. Authorization failed.
    Enter your PIN  (Attempt 2/3): 1221
    Incorrect PIN. Authorization failed.
    Enter your PIN  (Attempt 3/3): 1112
    Incorrect PIN. Authorization failed.
    Exiting

#Section 8: Additional If-Else Practice (30 Questions)
##Problem 8.1: Positive, Negative, or Zero (Easy)
Write a program that checks if a number is positive, negative, or zero.

Hint: Use if-elif-else: if n > 0 (positive), elif n < 0 (negative), else (zero).

In [3]:
print("Positive, Negative, or Zero")

my_number = int(input("Enter a number: "))

if my_number > 0:
  print(f"{my_number} is a positive number")

elif my_number < 0:
    print(f"{my_number} is a negative number")

else:
  print(f"{my_number} is zero")

Positive, Negative, or Zero
Enter a number: -2
-2 is a negative number


    Positive, Negative, or Zero
    Enter a number: 0
    0 is zero

    Positive, Negative, or Zero
    Enter a number: 9
    9 is a positive number

##Problem 8.2: Even or Odd (Easy)
Write a program that determines if a number is even or odd.

Hint: If n % 2 == 0, it's even; otherwise odd.

In [4]:
print("Even or Odd")

my_number = int(input("Enter a number: "))

if my_number % 2 == 0:
  print(f"{my_number} is an even number")

else:
    print(f"{my_number} is an odd number")

Even or Odd
Enter a number: 45
45 is an odd number


##Problem 8.3: Largest of Three Numbers (Easy)
Write a program to find the largest of three numbers.

Hint: Compare: if a >= b and a >= c, largest is a. Use nested if-else or use max() function.

In [8]:
print("Largest of Three Numbers")

a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
c = int(input("Enter third number: "))

if a >= b and a >= c:
    largest = a

elif b >= a and b >= c:
  largest = b

else:
  largest = c

print(f"The largest number is : {largest}")

Largest of Three Numbers
Enter first number: 34
Enter second number: 43
Enter third number: 12
The largest number is : 43
